File: Demos/Android_CS/MainActivity.cs

using SpaceFlint.Demos;

namespace com.spaceflint.bluebonnet.csharp
{

    public class MainActivity : android.app.Activity
    {

        protected override void onCreate(android.os.Bundle savedInstanceState)
        {
            base.onCreate(savedInstanceState);

            getWindow().setFlags(android.view.WindowManager.LayoutParams.FLAG_FULLSCREEN,
                                 android.view.WindowManager.LayoutParams.FLAG_FULLSCREEN);
            requestWindowFeature(android.view.Window.FEATURE_NO_TITLE);

            var canvasView = new CanvasView(this);
            setContentView(canvasView);
            canvasView.requestFocus();

            Points_CS.Initialize(canvasView);

            InitTimer(canvasView);
        }

        void InitTimer(CanvasView view)
        {
            const int TIMER_RATE = 16;
            java.lang.Runnable runnable = null;
            runnable = ((java.lang.Runnable.Delegate) (() =>
            {
                view.invalidate();
                view.postDelayed(runnable, TIMER_RATE);

            })).AsInterface();
            view.postDelayed(runnable, TIMER_RATE);
        }
    }

    public class CanvasView : android.view.View, HAL {

        System.Action frameCallback;
        android.graphics.Paint paint;
        android.graphics.Canvas frameCanvas;
        int frameWidth, frameHeight;

        public CanvasView(android.content.Context context) : base(context)
        {
            paint = new android.graphics.Paint();
            paint.setColor(android.graphics.Color.RED);
            paint.setAntiAlias(true);
            paint.setStrokeWidth(4);
            paint.setTextSize(30);
        }

        protected override void onDraw(android.graphics.Canvas canvas)
        {
            (frameCanvas, frameWidth, frameHeight) =
                (canvas, canvas.getWidth(), canvas.getHeight());

            paint.setColor(0x7FFF0000);
            drawCenter("Bluebonnet Demo", -4f);
            drawCenter("Android - C#", 4f);

            frameCallback();
        }

        void drawCenter(string text, float yMultiply)
        {
            var r = new android.graphics.Rect();
            frameCanvas.getClipBounds(r);
            paint.setTextAlign(android.graphics.Paint.Align.LEFT);
            paint.getTextBounds(text, 0, text.Length, r);
            float x = frameWidth / 2f - r.width() / 2f - r.left;
            float y = frameHeight / 2f + (r.height() * yMultiply) / 2f - r.bottom;
            frameCanvas.drawText(text, x, y, paint);
        }

        void HAL.Frame(System.Action callback) => frameCallback = callback;
        void HAL.Pixel(float x, float y, float r, float g, float b)
        {
            var x1 = Clamp(x, frameWidth);
            var y1 = Clamp(y, frameHeight);
            int r1 = Clamp(r, 256);
            int g1 = Clamp(g, 256);
            int b1 = Clamp(b, 256);
            paint.setColor((0xFF << 24) | (r1 << 16) | (g1 << 8) | b1);
            frameCanvas.drawPoint(x1, y1, paint);
            int Clamp(float a, int b) => System.Math.Min(System.Math.Max((int) (a * b), 0), b - 1);
        }
        float HAL.Random() => (float) java.lang.Math.random();

    }

}

File: Demos/Android_FS/MainActivity.fs

namespace com.spaceflint.bluebonnet.fsharp
open SpaceFlint.Demos

type MainActivity() =
    inherit android.app.Activity()

    let initTimer (view : android.view.View) =
        let mutable runnable : java.lang.Runnable = null
        runnable <- (new java.lang.Runnable.Delegate (fun () ->
                view.invalidate()
                view.postDelayed(runnable, 16L) |> ignore
            )).AsInterface()
        view.postDelayed(runnable, 16L) |> ignore
        ()

    override this.onCreate(savedInstanceState: android.os.Bundle) =
        base.onCreate(savedInstanceState)

        base.getWindow().setFlags(android.view.WindowManager.LayoutParams.FLAG_FULLSCREEN,
                                  android.view.WindowManager.LayoutParams.FLAG_FULLSCREEN)

        base.requestWindowFeature(android.view.Window.FEATURE_NO_TITLE) |> ignore

        let canvasView = new CanvasView(this)
        this.setContentView canvasView
        canvasView.requestFocus() |> ignore

        SpaceFlint.Demos.Points.myInitialize canvasView

        initTimer canvasView

and CanvasView(context) =
    inherit android.view.View(context)

    let paint = new android.graphics.Paint()
    let mutable frameCallback : System.Action = null
    let mutable frameCanvas : android.graphics.Canvas = null
    let mutable frameSize : (int * int) = (0, 0)

    do
        paint.setColor(android.graphics.Color.RED)
        paint.setAntiAlias(true)
        paint.setStrokeWidth(4.0f)
        paint.setTextSize(30.0f)

    let drawCenter text yMultiply =

        let r = new android.graphics.Rect()
        frameCanvas.getClipBounds(r) |> ignore
        paint.setTextAlign(android.graphics.Paint.Align.LEFT)
        paint.getTextBounds(string text, 0, (string text).Length, r)
        let (frameWidth, frameHeight) = frameSize
        let x = single frameWidth / 2.0f
              - single (r.width()) / 2.0f
              - single (r.left)
        let y = single frameHeight / 2.0f
              + single (r.height()) * single yMultiply / 2.0f
              - single r.bottom
        frameCanvas.drawText(text, x, y, paint)

    override this.onDraw(canvas) =

        frameCanvas <- canvas
        frameSize <- (canvas.getWidth(), canvas.getHeight())

        paint.setColor(0x7FFF0000)
        drawCenter "Bluebonnet Demo" -4.0f
        drawCenter "Android - F#"     4.0f

        frameCallback.Invoke()

    interface HAL with

        member this.Frame frameCallbackArg =

            frameCallback <- frameCallbackArg
            ()

        member this.Pixel (x, y, r, g, b) =

            let clamp (a:single) b = System.Math.Min(System.Math.Max(int (a * single b), 0), b - 1)
            let (frameWidth, frameHeight) = frameSize
            let x1 = clamp x frameWidth
            let y1 = clamp y frameHeight
            let r1 = clamp r 256
            let g1 = clamp g 256
            let b1 = clamp b 256
            paint.setColor((0xFF <<< 24) ||| (r1 <<< 16) ||| (g1 <<< 8) ||| b1)
            frameCanvas.drawPoint(single x1, single y1, paint)

        member this.Random () = single (java.lang.Math.random())
